### PR TITLE
Improve search performance (caching, cancellation, parallel search)

### PR DESCRIPTION
**Pull request by drognanar**
Improves search performance. Returns results **without any observable delay** when searching through large docsets (including .NET and QT).

Cancels previous incomplete searches.
Defers browser url loading.
Adds caching of previous searches. Filters through the existing result list to avoid searching through the entire docset.
Adds parallel searching of docsets.

Depends on #459 in order to correctly display ToC after ESC is pressed.

### CODE REVIEW

**Comment by trollixx** [comment]
This is an awesome work. Thank you!

I'll be going over and cherry-picking commits individually shortly. I wish GitHub could create series of PRs like Gerrit does.

File: src/registry/docset.cpp [hunk 1]
```diff
@@ -412,6 +472,7 @@ void Docset::countSymbols()
         const QString symbolType = parseSymbolType(symbolTypeStr);
         m_symbolStrings.insertMulti(symbolType, symbolTypeStr);
         m_symbolCounts[symbolType] += query.value(1).toInt();
+        m_symbolsTotal += query.value(1).toInt();
```
Comment (trollixx) [inline src/registry/docset.cpp]: I don't see this being used anywhere...

File: src/ui/mainwindow.cpp [hunk 2]
```diff
@@ -148,7 +149,7 @@ MainWindow::MainWindow(Core::Application *app, QWidget *parent) :
     connect(ui->actionCheckForUpdate, &QAction::triggered,
             m_application, &Core::Application::checkForUpdate);
     connect(ui->actionAboutZeal, &QAction::triggered, [this]() {
-        QScopedPointer<AboutDialog> dialog(new AboutDialog(this));
+        std::unique_ptr<AboutDialog> dialog(new AboutDialog(this));
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: I am really curious, apart from the support for move semantics that we don't use, what's the actual benefit from switching from `QScopedPointer` to `std::unique_ptr`?

File: src/ui/mainwindow.h [hunk 3]
```diff
@@ -134,17 +140,22 @@ private slots:
...
+    std::unique_ptr<Zeal::SettingsDialog> m_settingsDialog;
+
+    std::unique_ptr<QMenu> m_backMenu;
+    std::unique_ptr<QMenu> m_forwardMenu;
 
-    QMenu *m_backMenu = nullptr;
-    QMenu *m_forwardMenu = nullptr;
+    Zeal::CancellationToken m_cancelSearch;
 
+    std::unique_ptr<QShortcut> m_focusSearch;
```
Comment (trollixx) [inline src/ui/mainwindow.h]: I don't see how keeping this is useful.

File: src/ui/mainwindow.cpp [hunk 4]
```diff
@@ -99,9 +101,9 @@ MainWindow::MainWindow(Core::Application *app, QWidget *parent) :
     // initialise ui
     ui->setupUi(this);
 
-    QShortcut *focusSearch = new QShortcut(QKeySequence(QStringLiteral("Ctrl+K")), this);
-    focusSearch->setContext(Qt::ApplicationShortcut);
-    connect(focusSearch, &QShortcut::activated,
+    m_focusSearch = std::unique_ptr<QShortcut>(new QShortcut(QKeySequence(QStringLiteral("Ctrl+K")), this));
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: `QShortcut` is `QObject`, it is deleted when by its parent. Why `std::unique_ptr`?

File: src/ui/mainwindow.h [hunk 5]
```diff
@@ -134,17 +140,22 @@ private slots:
...
     SearchState *m_searchState = nullptr;
-    Zeal::NetworkAccessManager *m_zealNetworkManager = nullptr;
 
-    Ui::MainWindow *ui = nullptr;
+    std::unique_ptr<Ui::MainWindow> ui;
     Zeal::Core::Application *m_application = nullptr;
     Zeal::Core::Settings *m_settings = nullptr;
-    Zeal::ListModel *m_zealListModel = nullptr;
-    Zeal::SettingsDialog *m_settingsDialog = nullptr;
+    std::unique_ptr<Zeal::NetworkAccessManager> m_zealNetworkManager;
```
Comment (trollixx) [inline src/ui/mainwindow.h]: Same as with `QShortcut`, all these are `QObject`s with set parent, there's no need in any additional memory management.

File: src/registry/docsetsearchstrategy.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+/// This class contains search related methods for docsets.
+class DocsetSearchStrategy
+{
+public:
+    DocsetSearchStrategy();
+    virtual QList<SearchResult> search(const QString &query, CancellationToken token) = 0;
+
+    /// Used to filter out cached results.
+    virtual bool validResult(const QString &query, SearchResult previousResult,
```
Comment (trollixx) [inline src/registry/docsetsearchstrategy.h]: Why make a copy of `previousResult`?

File: src/registry/docsetsearchstrategy.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+namespace Zeal {
+
+class SearchResult;
+
+/// This class contains search related methods for docsets.
+class DocsetSearchStrategy
+{
+public:
+    DocsetSearchStrategy();
+    virtual QList<SearchResult> search(const QString &query, CancellationToken token) = 0;
```
Comment (trollixx) [inline src/registry/docsetsearchstrategy.h]: Why make a copy of `token`?

File: src/registry/docsetregistry.cpp [hunk 8]
```diff
@@ -121,25 +125,32 @@ void DocsetRegistry::_addDocset(const QString &path)
...
-
-    for (Docset *docset : docsets())
-        m_queryResults << docset->search(query);
+    finalResult += partial;
+}
 
-    std::sort(m_queryResults.begin(), m_queryResults.end());
+void DocsetRegistry::_runQueryAsync(const QString &query, const CancellationToken token)
+{
+    QFuture<QList<SearchResult> > queryResultsFuture = QtConcurrent::mappedReduced(
```
Comment (trollixx) [inline src/registry/docsetregistry.cpp]: I really like this. I think we could utilise `QFutureWatcher` here, and get rid of the `QMetaObject::invokeMethod()` and `m_thread`, right?

Also, no need in the space between `>>`, C++11 is required to build the project.

File: src/registry/cancellationtoken.h [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+#ifndef CANCELLATIONTOKEN_H
+#define CANCELLATIONTOKEN_H
+
+#include <QSharedPointer>
+
+namespace Zeal {
+
+/// Token that stores whether cancel was called on it.
+/// In async code can be used to check if another thread called cancel.
+struct CancellationToken
```
Comment (trollixx) [inline src/registry/cancellationtoken.h]: As we briefly discussed, it feels like we could use `QFutureWatcher::cancel()` instead of a custom solution. Have you got a chance to look into it?

Comment (drognanar) [inline src/registry/cancellationtoken.h]: I finally had the chance. `QFutureWatcher::cancel(); QFutureWatcher::waitForFinished()` waits until all existing threads finish and does not start new ones. Thus, worst case, `cancel` would take as much time as it takes to search the largest docset.
On the other hand `CancellationToken` allows to stop a search thread at any point during its execution.

File: src/registry/docsetregistry.cpp (hunk 8 above)
Comment (drognanar) [inline src/registry/docsetregistry.cpp]: It definitely is possible to get rid of `invokeMethod`, but I'd rather do that in a separate PR.
Also, I was used to writing `> >` and only realized that this is not necessary after pushing this PR.

**Comment by drognanar** [comment]
Made changes to the PR. I rolled back changes I made with regards to pointers.

File: src/registry/cachingsearchstrategy.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with Zeal. If not, see <http://www.gnu.org/licenses/>.
+**
+****************************************************************************/
+
+#include "cachingsearchstrategy.h"
+#include "docset.h"
```
Comment (trollixx) [inline src/registry/cachingsearchstrategy.cpp]: Please sort alphabetically, adding an empty line after `cachingsearchstrategy.h`.

File: src/registry/cachingsearchstrategy.cpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+#include "searchresult.h"
+#include "searchquery.h"
+
+using namespace Zeal;
+
+CachingSearchStrategy::CachingSearchStrategy(std::unique_ptr<DocsetSearchStrategy> strategy)
+    : m_search(std::move(strategy)),
```
Comment (trollixx) [inline src/registry/cachingsearchstrategy.cpp]: Zeal is currently uses other style, but I do think to switch over. Anyway, in this case comma should be on the next line.

File: src/registry/cachingsearchstrategy.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    QList<SearchResult> search(const QString &query, CancellationToken &token) override;
+    bool validResult(const QString &query, const SearchResult &previousResult,
+                     SearchResult &result)override;
+
+private:
+    QString getCacheEntry(const QString &query) const;
+    QList<SearchResult> searchWithCache(const QString &query, const QString &prefix);
+    QList<SearchResult> searchWithoutCache(const QString &query, CancellationToken token);
+
+    const static int CacheSize = 10;
```
Comment (trollixx) [inline src/registry/cachingsearchstrategy.h]: I think this should live in the .cpp file scope.

File: src/registry/cancellationtoken.cpp [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+using namespace Zeal;
+
+CancellationToken::CancellationToken()
+{
+    m_cancelled = QSharedPointer<bool>(new bool(false));
+}
+
+void CancellationToken::cancel()
+{
+    *m_cancelled = true;
```
Comment (trollixx) [inline src/registry/cancellationtoken.cpp]: This is not an atomic operation, but I am not sure if it can cause any problems...

File: src/registry/docsetregistry.cpp [hunk 14]
```diff
@@ -120,25 +124,32 @@ void DocsetRegistry::_addDocset(const QString &path)
...
 
-void DocsetRegistry::search(const QString &query)
+void DocsetRegistry::search(const QString &query, CancellationToken token)
 {
-    // Only invalidate queries
-    if (query.isEmpty())
-        return;
-
-    QMetaObject::invokeMethod(this, "_runQuery", Qt::QueuedConnection, Q_ARG(QString, query));
+    qRegisterMetaType<CancellationToken>("CancellationToken");
```
Comment (trollixx) [inline src/registry/docsetregistry.cpp]: Why register both?

File: src/registry/docsetregistry.cpp [hunk 15]
```diff
@@ -120,25 +124,32 @@ void DocsetRegistry::_addDocset(const QString &path)
...
+    finalResult += partial;
+}
 
-    std::sort(m_queryResults.begin(), m_queryResults.end());
+void DocsetRegistry::_runQueryAsync(const QString &query, const CancellationToken token)
+{
+    QFuture<QList<SearchResult> > queryResultsFuture = QtConcurrent::mappedReduced(
+                docsets(),
+                std::bind(&Docset::search, std::placeholders::_1, query, token),
+                &MergeQueryResults);
```
Comment (trollixx) [inline src/registry/docsetregistry.cpp]: I'd prefer a small lambda instead of `MergeQueryResults`.

File: src/registry/cancellationtoken.cpp (hunk 13 above)
Comment (drognanar) [inline src/registry/cancellationtoken.cpp]: There are no write races since mainwindow is the only place where the token gets cancelled.
And even if there would be many threads cancelling all set the value to true.
Finally, the shared pointer is not reassigned and cannot be disposed since at least one instance of CancellationToken exists (the one that is calling the cancel method).

File: src/registry/docsetregistry.cpp (hunk 14 above)
Comment (drognanar) [inline src/registry/docsetregistry.cpp]: I remember that I had issues when registering the meta type here earlier.
However, I no longer recall the exact problem and after removing one of the calls did not break zeal.

File: src/registry/docsetregistry.cpp (hunk 15 above)
Comment (drognanar) [inline src/registry/docsetregistry.cpp]: I tried but the program did not compile when lambda was passed in.

**Comment by drognanar** [comment]
Another update

**Comment by mrhota** [comment]
@drognanar Any updates here? I can rebase this if you don't have the time.

**Comment by mrhota** [comment]
@trollixx @drognanar what needs to happen to move this along? Fuzzy searching is waiting on this.

**Comment by trollixx** [comment]
I've been going over this multiple times, and the implementation details seem very counterintuitive to me. Exposing docset internals outside of `Docset` class is one of such things. It's quite opposite from what I was trying to achieve by encapsulating everything in one class.

I am really not comfortable with merging it right away, as I am not completely sold on search strategies idea. Chaining multiple is not a really flexible approach. For example, it leads to  iterating over results directly, while there's no way to apply custom search logic. Also the order in which strategies can be chained is also fixed, what I'd consider a design flaw.

There are a decent amount of unrelated changes here, like replacing `QScopedPointer` with `std::unique_ptr`, removing `Docset::path()`, etc.

I'd prefer to hold this off a bit longer, and see how's other changes from this PR will do in the next release.

**Comment by trollixx** [comment]
This is a very old and abandoned PR, let's put it to rest.
